File: src/javascript/population.js
let genomeInputsN = 2;
let genomeOutputN = 1;
let showBest = true;

class Population {
	constructor(size) {
		this.population = [];
		this.bestPlayer;
		this.bestFitness = 0;

		this.generation = 0;
		this.matingPool = [];

		for (let i = 0; i < size; i++) {
			this.population.push(new Player(i));
			this.population[i].brain.generateNetwork();
			this.population[i].brain.mutate();
		}
	}

	updateAlive() {
		for (let i = 0; i < this.population.length; i++) {
			const player = this.population[i];
			if (!player.dead) {
				player.look();
				player.think();
				player.move();
				player.update();
			}
		}

		if (showBest && this.bestPlayer && !this.bestPlayer.dead) {
			this.bestPlayer.look();
			this.bestPlayer.think();
			this.bestPlayer.move();
			this.bestPlayer.update();
			this.bestPlayer.show();
		}
	}

	done() {
		for (let i = 0; i < this.population.length; i++) {
			if (!this.population[i].dead) {
				return false;
			}
		}
		return true;
	}

	naturalSelection() {
		this.calculateFitness();

		let averageSum = this.getAverageScore();
		console.log(averageSum);
		let children = [];

		this.fillMatingPool();
		for (let i = 0; i < this.population.length; i++) {
			let parent1 = this.selectPlayer();
			let parent2 = this.selectPlayer();
			if (parent1.fitness > parent2.fitness) {
				children.push(parent1.crossover(parent2));
			} else {
				children.push(parent2.crossover(parent1));
			}
		}

		this.population.splice(0, this.population.length);
		this.population = children.slice(0);
		this.generation++;
		for (let i = 0; i < this.population.length; i++) {
			this.population[i].brain.generateNetwork();
		}

		console.log('Generation: ' + this.generation);

		this.bestPlayer.lifespan = 0;
		this.bestPlayer.dead = false;
		this.bestPlayer.score = 1;
	}

	calculateFitness() {
		let currentMax = 0;
		for (let i = 0; i < this.population.length; i++) {
			const player = this.population[i];
			player.calculateFitness();
			if (player.fitness > this.bestFitness) {
				this.bestFitness = player.fitness;
				this.bestPlayer = player.clone();
				this.bestPlayer.brain.id = 'BestGenome';
				this.bestPlayer.brain.draw();
			}

			if (player.fitness > currentMax) {
				currentMax = player.fitness;
			}
		}

		// normalize
		for (let i = 0; i < this.population.length; i++) {
			this.population[i].fitness /= currentMax;
		}
	}

	fillMatingPool() {
		this.matingPool.splice(0, this.matingPool.length);
		for (let i = 0; i < this.population.length; i++) {
			let n = this.population[i].fitness * 100;
			for (let j = 0; j < n; j++) {
				// ? why index only
				this.matingPool.push(i);
			}
		}
	}

	selectPlayer() {
		const rand = Math.floor(Math.random() * this.matingPool.length);
		return this.population[this.matingPool[rand]];
	}

	getAverageScore() {
		let averageSum = 0;
		for (let i = 0; i < this.population.length; i++) {
			averageSum += this.population[i].score;
		}
		return averageSum / this.population.length;
	}
}

File: src/cpp/population.cpp
#include <iostream>

#include "./header/population.h"
#include "./header/randomFun.h"
namespace neatCpp {
    Population::Population(int populationSize, int inputSize, int outputSize) {
        bestFitness = 0;
        generation = 0;
        bestPlayer = nullptr;
        for (int i = 0;i < populationSize;++i) {
            Player* p = new Player(i, inputSize, outputSize);
            p->getBrain()->generateNetwork();
            p->getBrain()->mutate();
            population.push_back(p);
        }
    }
    Population::~Population() {
        if (bestPlayer)
            delete bestPlayer;
        for (int i = 0; i < population.size();++i) {
            delete population[i];
        }
    }
    bool Population::populationDone() {
        for (int i = 0; i < population.size();++i) {
            if (!population[i]->isDead()) {
                return false;
            }
        }
        return true;
    }
    void Population::naturalSelection() {
        calculateFitness();
        long double averageSum = getAverageScore();
        std::cout << "Average Sum of Generation " << generation << ": " << averageSum << std::endl;
        std::vector<Player*> children;
        fillMatingPool();
        for (int i = 0; i < population.size(); ++i) {
            Player* parent1 = selectPlayer();
            Player* parent2 = selectPlayer();
            // removed compare, already in crossover
            children.push_back(parent1->crossover(parent2));

        }
        deleteAllPlayer();
        population = children;
        ++generation;
        //
        for (int i = 0;i < population.size();++i) {
            population[i]->getBrain()->generateNetwork();
        }
        std::cout << "Generation: " << generation << " created" << std::endl;
    }
    //
    void Population::calculateFitness() {
        long double currentMax = 0;
        for (int i = 0; i < population.size();++i) {
            Player* player = population[i];
            player->calculateFitness();
            long double fitness = player->getFitness();
            if (fitness > bestFitness) {
                bestFitness = fitness;
                if (bestPlayer) {
                    delete bestPlayer;
                    bestPlayer = nullptr;
                }
                bestPlayer = player->clone();
            }

            if (fitness > currentMax) {
                currentMax = fitness;
            }
        }
        // normalize
        for (int i = 0; i < population.size(); ++i) {
            population[i]->setFitness(population[i]->getFitness() / currentMax);
        }
    }
    void Population::fillMatingPool() {
        matingPool.clear();
        for (int i = 0; i < population.size();++i) {
            const long int n = population[i]->getFitness() * 100;
            for (int j = 0; j < n; ++j) {
                matingPool.push_back(i);
            }
        }
    }
    Player* Population::selectPlayer() {
        return population[matingPool[(int)(randNum(0, matingPool.size()))]];
    }
    long double Population::getAverageScore() {
        long int averageSum = 0;
        for (int i = 0; i < population.size();++i) {
            averageSum += population[i]->getScore();
        }
        return averageSum / population.size();
    }
    void Population::deleteAllPlayer() {
        for (int i = 0; i < population.size(); ++i) {
            delete population[i];
            population[i] = nullptr;
        }
    }
}